eclair-core/src/main/scala/fr/acinq/eclair/payment/MultiPartPaymentLifecycle.scala

```diff
@@ -0,0 +1,330 @@
+/*
+ * Copyright 2019 ACINQ SAS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.payment
+
+import java.util.UUID
+
+import akka.actor.{ActorContext, ActorRef, FSM, Props}
+import akka.event.Logging.MDC
+import fr.acinq.bitcoin.ByteVector32
+import fr.acinq.eclair.db.{OutgoingPayment, OutgoingPaymentStatus, PaymentsDb}
+import fr.acinq.eclair.payment.PaymentInitiator.SendPaymentRequest
+import fr.acinq.eclair.payment.PaymentLifecycle.{PaymentProgressHandler, SendPayment}
+import fr.acinq.eclair.payment.PaymentSent.PartialPayment
+import fr.acinq.eclair.router._
+import fr.acinq.eclair.transactions.Transactions
+import fr.acinq.eclair.wire.Onion
+import fr.acinq.eclair.{FSMDiagnosticActorLogging, Logs, LongToBtcAmount, MilliSatoshi, NodeParams, ToMilliSatoshiConversion}
+
+import scala.compat.Platform
+import scala.util.Random
+
+/**
+ * Created by t-bast on 18/07/2019.
+ */
+
+/**
+ * Sender for a multi-part payment (see https://github.com/lightningnetwork/lightning-rfc/blob/master/04-onion-routing.md#basic-multi-part-payments).
+ * The payment will be split into multiple sub-payments that will be sent in parallel.
+ */
+class MultiPartPaymentLifecycle(nodeParams: NodeParams, id: UUID, relayer: ActorRef, router: ActorRef, register: ActorRef) extends FSMDiagnosticActorLogging[MultiPartPaymentLifecycle.State, MultiPartPaymentLifecycle.Data] {
+
+  import MultiPartPaymentLifecycle._
+
+  val paymentsDb = nodeParams.db.payments
+
+  startWith(PAYMENT_INIT, PaymentInit(None, None, None))
+
+  when(PAYMENT_INIT) {
+    case Event(r: SendPaymentRequest, d: PaymentInit) =>
+      require(d.request.isEmpty, "multi-part payment lifecycle must receive only one payment request")
+      router ! GetNetworkStats
+      stay using d.copy(sender = Some(sender()), request = Some(r))
+
+    case Event(s: GetNetworkStatsResponse, d: PaymentInit) =>
+      require(d.request.nonEmpty && d.sender.nonEmpty, "multi-part payment request must be set")
+      val r = d.request.get
+      s.stats match {
+        case Some(networkStats) =>
+          log.debug(s"network stats available: ${networkStats.capacity}")
+          relayer ! GetUsableBalances
+          goto(PAYMENT_IN_PROGRESS) using PaymentProgress(d.sender.get, r, networkStats, r.amount, r.maxAttempts, Map.empty, Nil)
+        case None =>
+          log.debug("network stats not available: retrying...")
+          router ! TickComputeNetworkStats
```

Comment:
It's okay-ish to attempt the payment without network stats, but it's likely better to do it with accurate network stats.

I updated the code to do without network stats if we don't have them, but trigger a computation on the router for potential next payments. But I think we can improve when network stats are computed, and I'm interested in your feedback on that. Right now I only compute network stats when we finish a sync (and I don't compute them from the "old" public channels data we had in the DB). Maybe it would make more sense to compute a first version of the network stats when the router boots from what's inside the DB, and then re-compute it once a sync completes?
If we do that maybe we don't need the `MultiPartPaymentLifecycle` to tell the router to re-compute?

Reply:
See https://github.com/ACINQ/eclair/pull/1202/commits/0d019864a778b934eb0b5c8dbfc77b92f1c7fe26